rfcs/0189-contracts.md

```diff
@@ -0,0 +1,658 @@
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+## Dual Link
+[unresolved-duallink]: #unresolved-questions-duallink
+
+The current implementation requires the `end user` to link the consumer and provider
+in both directions:
+
+```nix
+config = {
+  # consumer to provider
+  services.nextcloud.fileBackup.provider = config.services.restic.backups.nextcloud.fileBackup;
+
+  # provider to consumer
+  services.restic.backups.nextcloud.fileBackup.consumer = config.services.nextcloud.fileBackup;
+};
+```
+
+It would be so much nicer if we could somehow require only to give the `consumer` to the `provider`
+and it managed to make the link backwards automatically.
+In the snippet above, this means removing the need for the `provider to consumer` line.
+
+The issue comes from the `consumer` and `provider` option in the top-level `contracts` definition to be of type `optionType`.
+They don't have access to the actual `input` and `output` values of an instantiated contract.
+
+Experimenting on this has been done in the [module interfaces][] repo.
+There, we set the `provider` option as a function which takes an argument
+which is the instantiated `consumer`, so it is not of type `optionType` but of type `submodule` and has access to the real input and output values.
+Unfortunately, this has two downsides:
+
+1. It requires one more line in each provider definition. This would be okay if there wasn't the following downside.
+2. There's no way to write side effects. This means the `provider` can only write to its own `output`, which misses the whole point of having contracts in the first place.
+
+There's maybe a way to solve this but we didn't figure it out. Help is appreciated!
+Beware though you will be crossing the edge of the module system and entering the land of infinite recursion.
+
+## Documentation
+[unresolved-documentation]: #unresolved-questions-documentation
+
+It is not possible to build the manual right now. Doing so results in an error.
+
+```bash
+$ (cd nixos/; nix-build release.nix -A manual.x86_64-linux)
+
+[...]
+
+       error: attribute 'contracts' missing
+       at /home/timi/Projects/nixpkgs/nixos/modules/services/web-apps/stash.nix:435:16:
+          434|       jwtSecretKeyFile = mkOption {
+          435|         type = config.contracts.secret.consumer;
+             |                ^
+          436|         description = "Path to file containing a secret used to sign JWT tokens.";
+```
+
+Comments in the [draft PR][draftPR] have been added to indicate what has been tried.
+Any help is appreciated to solve this.
+
+# Future work
+[future]: #future-work
+
+- Solve the [documentation][unresolved-documentation] issue.
+- Identify contracts and their inputs, outputs and behavior tests.
+- Identify services that would benefit from being consumers and providers of contracts and add the necessary options.
+- Optionally solve the [dual-link][unresolved-duallink] issue.
```

Comment:
clicking these links seems not to actually navigate the user to these on the page i think it wants the ids instead?

Reply:
Just checked it is fixed now :)